include/appbase/application.hpp

```diff
@@ -1,383 +1,24 @@
-               return *method_type::get_method(itr->second);
-            } else {
-               methods.emplace(std::make_pair(key, method_type::make_unique()));
-               return  *method_type::get_method(methods.at(key));
-            }
-         }
-
-         /**
-          * Fetch a reference to the channel declared by the passed in type.  This will construct the channel
-          * on first access.  This allows loose and deferred binding between plugins
-          *
-          * @tparam ChannelDecl - @ref appbase::channel_decl
-          * @return reference to the channel described by the declaration
-          */
-         template<typename ChannelDecl>
-         auto get_channel() -> std::enable_if_t<is_channel_decl<ChannelDecl>::value, typename ChannelDecl::channel_type&>
-         {
-            using channel_type = typename ChannelDecl::channel_type;
-            auto key = std::type_index(typeid(ChannelDecl));
-            auto itr = channels.find(key);
-            if(itr != channels.end()) {
-               return *channel_type::get_channel(itr->second);
-            } else {
-               channels.emplace(std::make_pair(key, channel_type::make_unique()));
-               return  *channel_type::get_channel(channels.at(key));
-            }
-         }
-
-         /**
-          * Do not run io_service in any other threads, as application assumes single-threaded execution in exec().
-          * @return io_serivice of application
-          */
-         boost::asio::io_service& get_io_service() { return io_serv; }
-
-         /**
-          * Post func to run on io_service with given priority.
-          *
-          * @param priority can be appbase::priority::* constants or any int, larger ints run first
-          * @param func function to run on io_service
-          * @return result of boost::asio::post
-          */
-         template <typename Func>
-         auto post( int priority, Func&& func ) {
-            return boost::asio::post(io_serv, pri_queue.wrap(priority, std::forward<Func>(func)));
-         }
-
-         /**
-          * Provide access to execution priority queue so it can be used to wrap functions for
-          * prioritized execution.
-          *
-          * Example:
-          *   boost::asio::steady_timer timer( app().get_io_service() );
-          *   timer.async_wait( app().get_priority_queue().wrap(priority::low, [](){ do_something(); }) );
-          */
-         auto& get_priority_queue() {
-            return pri_queue;
-         }
+// ---------------------------------------------------------------------------------------------
+// to use a different executor:
+//
+//   1. create a copy of this file in your own application
+//
+//   2. create your own version of appbase/default_executor.hpp
+//           (it needs to define a type `appbase_executor`)
+//      and include this file instead of <appbase/default_executor.hpp>
+//
+//   3. include your own `application.hpp` in your project
```

Comment:
This seems like too much of a hack to me. 

Seems like a user of appbase should choose which executor type either by a #define or better yet a type declaration before including `application.hpp`. I don't think we want users to have to hack up their copy of `appbase` to make it work.

Reply:
How to use `appbase`:

- Include the executor type you want (or write your own).
- Declare your executor as the one to use:
   ```
      #include <appbase/default_executor.hpp>
      namespace appbase { using appbase_executor = appbase::default_executor; }
      #include <appbase/application.hpp>
   ```

Reply:
Thanks Kevin. In leap for example, this pattern would have to be present in multiple files, right (wherever `#include <appbase/application.h>` is currently found). I don't think it would be good to repeat this patterm, but that we should have a single header for it. My suggestion was something like `<leap/application.h>`, which was the reason of my suggestion.
What would you suggest we do in appbase? Add the 3 lines above wherever `#include <appbase/application.h>` is currently found)?

Reply:
`appbase` is currently only used by plugins. So an `application.hpp` in `plugins/chain_interface/include/eosio/chain/` is likely a good place for it.

Reply:
> What would you suggest we do in appbase? Add the 3 lines above wherever #include <appbase/application.h> is currently found)?

Is there a problem with just making `application` a template? Then users could just do something like:
``` 
   #include <appbase/default_executor.hpp>
   #include <appbase/application.hpp>
   using app = appbase::application<default_executor>;
```

Reply:
Both Lin and I looked at making the whole `application` a template (that was the first idea), but that would have required moving all the functions and the includes that are currently inside `application_base.cpp` into the header file, and we both felt that it was not a good idea and that the current solution is much preferable.

Reply:
You should be able to keep application_base.cpp as is and only make `application` a template.

Reply:
> You should be able to keep application_base.cpp as is and only make `application` a template.

I don't see what problem that solves?

Reply:
Maybe we should have a quick meeting. My main point is that this all seems like a hack. A user of `appbase` should be able to use it directly without having to modify anything in the `appbase` submodule. We should provide both a `default_executor` and a `multi-queue-executor` and users should be able to choose between them. Or we just make our fork of `appbase` only have a multi-queue-executor since that is all we need. Not sure we need to maintain the `default_executor`. Seems like you should be able to use the multi-queue one in a way that you just only use one of the queues giving you the current default behavior.

Reply:
> A user of appbase should be able to use it directly without having to modify anything in the appbase submodule.

Yes, that is the case with the current implementation. It can be used without requiring any change in the appbase submodule.

Reply:
We can certainly have a quick meeting!

Reply:
We can provide a two-queue executor. The user can choose to post to either queue; by default the first queue is posted. We will have 2 execution modes: `from_first_queue` or `from_both_queues` and the user is able to set the mode either at the start or dynamically. I just thought those concepts were too specific for Leap. But now I think we should do this.

Reply:
Why would we do this? The whole point of the separation is that we want to allow the executor to be anything the user wants. Why hardcode two separate executors in appbase?

Reply:
I think we just change appbase to have only the new multi-queue executor. If you only need the one queue only post to the one queue.

Reply:
I will abstract what I already have for Leap and submit a default two-queue executor (it requires a little more changes like managing the order among two queues and support multiple threading). Should be done soon after lunch.

Reply:
> I think we just change appbase to have only the new multi-queue executor. 

I disagree. The executor used in leap will require more changes anyways, like having a thread_pool, etc... The point of this exercise was to *not* implement the needs of leap in appbase, so that leap can evolve without requiring changes to appbase.

Reply:
The users are free to implement their own executors. Appbase will just provide and maintain a default two-queue executor (thread-pool is not needed here). This two-queue executor still tries to be as generic as possible. Let me get it out and we will decide whether to maintain a one-queue or two-queue executor afterwards.

Reply:
I don't agree that  the two queue executor should be added in appbase, except maybe as an example. It adds a lot of unwarranted comp-lexity, and it is unlikely to be useful to any application except than leap. And as for leap, it will define its own executor anyways.

Reply:
The way I see it, the `default_executor` and the `execution_priority_queue` are examples which should be as basic as possible. 

The `priority_queue` is only needed by the executor.  A complex application with specific needs would redefine its own `executor` and `priority_queue` (or whatever enqueueing mechanism needed by the executor).